### ISSUE TITLE
Sersic profile created with very high count rates which potentially do not match the equivalent magnitude of the galaxy

### ISSUE THREAD

**Issue by Onoddil** [spacetelescope/STScI-STIPS#23]
Running STIPS I encounter an issue where the galaxies created have abnormally high countrates; a galaxy with a surface brightness of ~21.8 mag/arcsec^2 at z=0.3, with radius=0.7 arcsec results in STmag=21.2 (sensible, given m = S - 2.5 log10(A)), but a countrate of 101,000 counts/s. This then gives the resulting FITS file counts of >1e7 counts in a 1000 second exposure. Attached at the bottom of this issue are the files created during the process: the three internal galaxy files and the output log. The code used to create the files is also at the bottom of the page. As far as I have been able to trace the issue I believe it to lie in one of two possible places. 

1) The first is in ``instrument.py``, during ``readBC95Table`` where the BC95 .fits files are normalised to a given bandpass [here](https://github.com/spacetelescope/STScI-STIPS/blob/1fc072be5df097363247e86ad63b0e9b0c66124a/stips/instruments/instrument.py#L586-L599), which calls ``sp = self.normalize(sp, mag, norm_bp)`` with ``normalize`` creating a ``pandeia`` ``NormalizationFactory`` instance

```python
def normalize(self, source_spectrum_or_wave_flux, norm_flux, bandpass):
        from pandeia.engine.normalization import NormalizationFactory
        norm_type = self.get_type(bandpass)
        
        norm = NormalizationFactory(type=norm_type, bandpass=bandpass, norm_fluxunit='abmag', norm_flux=norm_flux)
```
however, ``norm_flux`` is what used to be ``mag`` in ``readBC95Table``, which is an iterable over [vmags](https://github.com/spacetelescope/STScI-STIPS/blob/1fc072be5df097363247e86ad63b0e9b0c66124a/stips/instruments/instrument.py#L576-L586), which is ``table['apparent_surface_brightness']``. Thus a surface brightness, in magnitudes per square arcseconds, is potentially being normalised to magnitudes via ``norm_fluxunit='abmag'``, as ``pandeia`` accepts several magnitudes and variations on Jansky as ``norm_fluxunit`` and does not offer any surface brightness normalisation.

2) The second potential issue may be in ``astro_image.py``, in ``addSersicProfile``, where ``flux`` is accepted as an input and used to [define](https://github.com/spacetelescope/STScI-STIPS/blob/1fc072be5df097363247e86ad63b0e9b0c66124a/stips/astro_image/astro_image.py#L485-L504) the ``amplitude`` of the ``astropy.modeling.models.Sersic2D`` instance. This is called from ``addTable``, where ``flux`` is [taken](https://github.com/spacetelescope/STScI-STIPS/blob/1fc072be5df097363247e86ad63b0e9b0c66124a/stips/astro_image/astro_image.py#L334-L410) from the internal file ``sim_gals_000_01_conv_Z087`` created by ``readBC95Table``. At this point the flux is still reasonable, but I believe is the _total galaxy flux_, not the amplitude of the profile at the effective radius. Thus when the image is defined via

```python
mod = Sersic2D(amplitude=pixel_brightness, r_eff=pixel_radius, n=n, x_0=xc, y_0=yc, ellip=(1.-axialRatio), theta=(np.radians(phi) + 0.5*np.pi))
img = mod(x, y)
```
it is wrong by a factor of the conversion from total profile flux to Ie. I believe the conversion could potentially be found [here](https://ned.ipac.caltech.edu/level5/March05/Graham/Graham2.html) and implemented along the lines of 

```python
from scipy.special import gamma, gammainc
bn = 2*n - 1/3
x_ = bn * (10**(1/n))  # set this to be 10*re away, which should be enough
g = gamma(2*n) * gammainc(2*n, x_)
i_e = flux / re**2 / 2 / np.pi / n / np.exp(bn) * bn**(2*n) / g
pixel_brightness = flux / (self.oversample*self.oversample)
pixel_brightness = pixel_brightness / flux * i_e
```
but I am unsure of how the profile is defined within STIPS, to know if reversing the integral is the correct approach. 
The problem could potentially be related to the transformation from detector to undersampled pixels, as the transformation from the detector pixels for defining the profile into the oversampled pixel grid (i.e., a position that is 2 pixels away suddenly becomes 10 pixels away for a 5x oversampling, with a corresponding 5x increase in the effective (pixel) radius) could lead to a difference in the values of the profile; the integral inside a circle of radius re (the effective radius) changes when [re = 10](https://www.wolframalpha.com/input/?i=integrate+r+exp(-(2-1%2F3)+*+((r%2F10)+-+1))+dr+from+0+to+10) compared to [re = 50](https://www.wolframalpha.com/input/?i=integrate+r+exp(-(2-1%2F3)+*+((r%2F50)+-+1))+dr+from+0+to+50) by a factor of the square of the change in re, in this instance a factor of 25. Thus the solution may lie in re-defining the pixels in undersampled but fractional pixels (i.e., pixel 0 becomes 0, 0.25, 0.5, 0.75, etc.), but again I am unsure of whether this is correct or not.
However, at this point the values for the Sersic profile (by default a one-second exposure) are very large, e.g.:
```
2018-11-01 14:16:00,193 INFO: Sersic profile has final size 132x132, maximum value 8769.47389331, sum 192943.784146
```
with ``central_flux`` being passed back to ``addTable``. Although ``central_flux`` is only the flux within the [half-light radius](https://github.com/spacetelescope/STScI-STIPS/blob/1fc072be5df097363247e86ad63b0e9b0c66124a/stips/astro_image/astro_image.py#L546-L548), and thus only ever providing half of the flux of the galaxy, this value is still much too large by this point, given that the note added to ``sim_gals_000_01_conv_Z087_WFIRST-WFI.txt`` is ``notes[item_index] = "{}: surface brightness {:.3f} yielded flux {:.3f}".format(notes[item_index], flux, central_flux)``, or ``BC95_a_10E9_21.837701: surface brightness 213.734 yielded flux 101110.321``. Here 'surface brightness' is ``flux`` which has units counts/s in ``sim_gals_000_01_conv_Z087.txt``, as it is derived in [instrument.py](https://github.com/spacetelescope/STScI-STIPS/blob/1fc072be5df097363247e86ad63b0e9b0c66124a/stips/instruments/instrument.py#L594-L595) as ``rate = obs.countrate()``.

In summary, there is an issue within STIPS where count rates for galaxies created by Sersic profiles result in unphysically high fluxes based on the quoted surface brightnesses and final magnitudes, potentially caused by a unit conversion issue. Culprits could include:
- ``NormalizationFactory`` conversion assuming a 'regular' ``ABmag`` magnitude and being given a surface brightness from ``sim_gals_000``
- The overall galaxy flux (in counts/s) being used as the effective radius amplitude of the ``astropy.modeling.models.Sersic2D`` profile
- Conversion from detector to oversampled pixels may change the normalisation of the Sersic profile
- Minor issues with output text files being inconsistent with magnitude, flux, count rate and surface brightness

Any help with this issue, or explanation of where I went wrong and how to fix the problem, would be greatly appreciated.

[galaxy_creation_test.log](https://github.com/spacetelescope/STScI-STIPS/files/2539564/galaxy_creation_test.log)
[sim_gals_000_01_conv_Z087_observed_WFIRST-WFI.txt](https://github.com/spacetelescope/STScI-STIPS/files/2539565/sim_gals_000_01_conv_Z087_observed_WFIRST-WFI.txt)
[sim_gals_000_01_conv_Z087.txt](https://github.com/spacetelescope/STScI-STIPS/files/2539566/sim_gals_000_01_conv_Z087.txt)
[sim_gals_000.txt](https://github.com/spacetelescope/STScI-STIPS/files/2539567/sim_gals_000.txt)

```python
import logging
from stips.scene_module import SceneModule
from stips.observation_module import ObservationModule
file_ = open('galaxy_creation_test.log', 'w+')
stream_handler = logging.StreamHandler(file_)
stream_handler.setFormatter(logging.Formatter('%(asctime)s %(levelname)s: %(message)s [in %(pathname)s:%(lineno)d]'))
logger = logging.getLogger()
logger.setLevel(logging.INFO)
logger.addHandler(stream_handler)

scm = SceneModule(logger=logger)

seedg = 9999
galaxy = {'n_gals': 50,
      'z_low': 0.0, 'z_high': 1.0,
      'rad_low': 0.5, 'rad_high': 2.0,
      'sb_v_low': 23.0, 'sb_v_high': 20.0,
      'distribution': 'uniform', 'clustered': False,
      'radius': 5, 'radius_units': 'arcsec',
      'offset_ra': 0.00001, 'offset_dec': 0.00001, 'seed': seedg}
galaxy_cat_file = scm.CreateGalaxies(galaxy)
seedo = 12345
obs = {'instrument': 'WFI',
       'filters': ['Z087'],
       'detectors': 1,
       'distortion': False,
       'oversample': 1,
       'pupil_mask': '',
       'background': 'avg',
       'observations_id': 1,
       'exptime': 1000,
       'offsets': [{'offset_id': 1, 'offset_centre': False, 'offset_ra': 0.0, 'offset_dec': 0.0, 'offset_pa': 0.0}],
       'small_subarray': True, 'seed': seedo}

obm = ObservationModule(obs, logger=logger)
obm.nextObservation()

output_galaxy_catalogues = obm.addCatalogue(galaxy_cat_file)
psf_file = obm.addError()
fits_file, mosaic_file, params = obm.finalize(mosaic=False)
```

**Comment by york-stsci** [comment]
Is anyone using STIPS release 1.0.7 still experiencing this issue? If not, I'll close it with the next STIPS release.

**Comment by robelgeda** [comment]
This seems like it has been resolved by v1.0.7, closing issue.
